File: Dynamic Programming/subsetsum.py
def SubsetSum(arr,n, s) : 
    if (s == 0) : 
        return True
    if (n == 0 and s != 0) : 
        return False
        
    if (arr[n - 1] > s) : 
        return SubsetSum(arr, n - 1, s); 
   
    return SubsetSum(arr, n-1, s) or SubsetSum(arr, n-1, s-arr[n-1]) 
t=int(input())
while(t>0):
    n=int(input())
    arr=list(map(int,input().split()))
    s=int(input())
    if(SubsetSum(arr,n,s)):
        print("YES")
    else:
        print("NO")
    t=t-1

File: Dynamic Programming/subsetsum.cpp
// Subset sum problem using Memoization
#include <iostream>
using namespace std;

// Return true if there exists a subarray of array[0..n) with given sum
bool subsetSum(int arr[], int n, int sum)
{
	// T[i][j] stores true if subset with sum j can be attained with
	// using items up to first i items
	bool T[n + 1][sum + 1];

	// if 0 items in the list and sum is non-zero
	for (int j = 1; j <= sum; j++)
		T[0][j] = false;

	// if sum is zero
	for (int i = 0; i <= n; i++)
		T[i][0] = true;

	// do for ith item
	for (int i = 1; i <= n; i++)
	{
		// consider all sum from 1 to sum
		for (int j = 1; j <= sum; j++)
		{
			// don't include ith element if j-arr[i-1] is negative
			if (arr[i - 1] > j)
				T[i][j] = T[i - 1][j];
			else
			// find subset with sum j by excluding or including the ith item
				T[i][j] = T[i - 1][j] || T[i - 1][j - arr[i - 1]];
		}
	}

	// return maximum value
	return T[n][sum];
}

// Subset Sum Problem
int main()
{
	// Input: set of items and a sum
	int arr[] = { 7, 3, 2, 5, 8 };
	int sum = 18;

	// number of items
	int n = sizeof(arr) / sizeof(arr[0]);

	if (subsetSum(arr, n, sum))
		cout << "Yes";
	else
		cout << "No";

	return 0;
}
